src/Report/Lcov.php

```diff
@@ -0,0 +1,122 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of phpunit/php-code-coverage.
+ *
+ * (c) Sebastian Bergmann <sebastian@phpunit.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+namespace SebastianBergmann\CodeCoverage\Report;
+
+use function count;
+use function dirname;
+use function file_put_contents;
+use function ksort;
+use function max;
+use function range;
+use SebastianBergmann\CodeCoverage\CodeCoverage;
+use SebastianBergmann\CodeCoverage\Directory;
+use SebastianBergmann\CodeCoverage\Driver\WriteOperationFailedException;
+use SebastianBergmann\CodeCoverage\Node\File;
+
+/**
+ * Ref: http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php.
+ */
+final class Lcov
+{
+    /**
+     * @throws WriteOperationFailedException
+     */
+    public function process(CodeCoverage $coverage, ?string $target = null, ?string $name = null): string
+    {
+        $lcovLines = [];
+        $report    = $coverage->getReport();
+
+        foreach ($report as $file) {
```

Comment:
My reading of the LCOV docs suggests that the final `.info` file should ideally not be simply organised by filename like this, but should be organised by test instead - i.e. for each testcase executed there should be a separate `TN:[testname]...end_of_record` block, with each block recording the hit data for the individual test in question.

Without doing that the generated HTML won't (can't) include a per-test breakdown like this one from the docs? http://ltp.sourceforge.net/coverage/lcov/output/example/methods/index-detail.html

Reply:
To be clearer, my reading of the `.info` spec suggests something like.

```
foreach ($tests as $test) {
    //TN:
    foreach($files as $file) {
        //SF: 
        // ....
        //end_of_record
    }
}
```

Reply:
Here's an example I got from the lcov test suite [target.txt](https://github.com/sebastianbergmann/php-code-coverage/files/7270454/target.txt)

Reply:
Thank you for taking the time to review this, I will reorganize it on a per test basis and investigate branches further :+1:

Reply:
Thank you, @dvdoug, for reviewing this in detail.

Reply:
@dvdoug I tried to wrap my head around refactoring to make it on a test basis, but I feel like `RawCodeCoverageData` is an aggregation of all the tests to source files and thus cannot split the data differently. It would require too much changes...

While I was adding the branch coverage, I noticed that the stats for the functions were off, `--coverage-html` vs `--coverage-lcov` + `genhtml`. But I think it is due to the renders, how they consider if a function is tested or not, one seem to not consider them tested if they are partially tested and the other does... The count for the lines and branches matched.

To note, to have `genhtml` generate the html for the branches you need to add `--branch-coverage`, and of course on the phpunit side needs to add `--path-coverage`.

Reply:
>I tried to wrap my head around refactoring to make it on a test basis, but I feel like RawCodeCoverageData is an aggregation of all the tests to source files and thus cannot split the data differently. It would require too much changes...

If you look at the built-in HTML report for a file and hover over a line, there's a tooltip produced that lists the individual test cases affecting that line so the data is available. Although the data is initially organised by file it should be possible to interrogate it to get what you need. On the basis that the order of output into `.info` doesn't matter, something like this?

```
foreach ($report as $file) {
   $tests = $this->getCoverageByTestcase($file);
   foreach ($tests as $test) {
      //TN: [$test]
      //SF: [$file]
      //...
   }
}
```

For the preprocessing, I think something like this should work:
```
private function getCoverageByTestcase($file) {
    $tests = array_keys($report->testData());
    foreach ($tests as $test) {
      $testSpecificData['line'][$test] = array_filter($report->lineCoverageData(), function($test) {...});
      $testSpecificData['branch'][$test] = array_filter($report->functionCoverageData(), function($test) {...});
    }
    return $testSpecificData; 
}

```

Reply:
@dvdoug I will investigate further :detective:

Reply:
@dvdoug Stalling again, I fear it is as I though.

The function `SebastianBergmann\CodeCoverage\Node::lineCoverageData` gives you something like this:
```
string(88) "SebastianBergmann\Environment\RuntimeTest::testGetVersionReturnsPhpVersionWhenRunningPhp"
197 => Array
(
    [0] => SebastianBergmann\Environment\RuntimeTest::testNameAndVersionCanBeRetrieved
    [1] => SebastianBergmann\Environment\RuntimeTest::testNameAndCodeCoverageDriverCanBeRetrieved
    [2] => SebastianBergmann\Environment\RuntimeTest::testGetVersionReturnsPhpVersionWhenRunningPhp
)
```
You have the line number as an index, but the test name will only appear once thus its impossible to tell how many times this specific test case hit that line. Has you can see in the popover in the built-in HTML coverage, it list how many test cases hit the line but **not how many times per test cases**. Which was an indicator to me that the data is just not there.

If organized by test cases in `lcov.info` we cant always put the `total hits` for all test cases lines, it has to be separated into each test cases.
![lcov-test-coverage](https://user-images.githubusercontent.com/748824/135746383-45ab79e8-6509-4781-8096-a5fc944d3166.png)

Unless you can point me to something I missed, I think what is currently there is the best that we can achieve with the data we have.

Reply:
>thus its impossible to tell how many times this specific test case hit that line

Are you sure that's what hit means in this context? I read it as 6 out of 8 lines in `iterate.c` were hit (at least once) by `test_2_to_2000`, 7 out of 8 lines were hit (at least once) by `test_overflow` etc. If it were intended to represent just a raw count (e.g. this test case hit the same line 5 times), it wouldn't have the denominator? From the example screenshot the denominators for the individual files (4 + 8) equal the total number of lines (12) displayed at top-right. Same for functions (1 + 1 = 2) and branches (2 + 4 = 6)

Reply:
You are right, wrong example :sweat_smile: . But that is just the rendered doing its thing believe. The fact is we have to provide:
```
DA:<line number>,<execution count>[,<checksum>]
```
**On a test case basis**. And that means we need to know how many time a specific test has hit a line. That is **if** we organize by test cases. If we dont group by test cases (_which is supported in `lcov` format, since it renders_) we wont be getting that extra section for each file. Which isnt a big deal as its meant mostly for external code coverage tools such as coveralls.io which have their github action only type `lcov` format.

Reply:
> \<execution count\>

That's unfortunate, neither Xdebug or PCOV collect that data so it's unfortunately not a question of changing php-code-coverage to pass it through.

However, I'm afraid I don't get this bit

> If we dont group by test cases (which is supported in lcov format, since it renders) we wont be getting that extra section for each file.

Even if the file were to be organised into one massive unnamed "test" (`TN: <blank>`), you'd still need to include `DA` segments for each because that's the only part that indicates whether an individual line is covered or not. Without it, you'd have no way of telling which lines aren't covered by tests (I note that in fact you are). What you're doing is coalescing the tests together and substituting the number of them that cover a line as the "execution count" as being the "best-available data".

I'm of the opinion that the best-fit to the format is to break it out by test, using a simple 0 or 1 for execution count if that's the best that can be done. For me, that keeps more granularity. It's always possible for LCOV-ingesting tools to combine the data for consumers who don't care about per-test breakdowns, but it's not possible the other way. For the purpose of producing summary statistics, my per-test 0/1 and your <count of tests> would in fact produce the same numbers.

One for @sebastianbergmann to decide!

Reply:
> I'm of the opinion that the best-fit to the format is to break it out by test, using a simple 0 or 1 for execution count if that's the best that can be done.

Makes sense to me.

Reply:
I'll re-organize it on a test basis and use `0, 1` for line hits :+1: 
Cheers! :partying_face: